sql/schemachanger: split computed column dependency rule to fix DROP COLUMN CASCADE deadlock

spilchen opened the pull request:
Previously, a single dependency rule applied to all computed column expressions, causing deadlocks when dropping columns with STORED computed dependencies via CASCADE. The rule forced computed expressions to be dropped before their referenced columns, but this created circular dependencies during CASCADE operations with STORED columns.

This change splits the rule into two separate rules:
- Virtual computed columns: Keep existing precedence rule needed for optimizer
- Stored computed columns: Only apply rule during ALTER COLUMN TYPE operations

This prevents the deadlock while preserving the necessary dependency ordering for ALTER COLUMN TYPE operations where temporary expressions map old rows to new column types.

Also adds NOTICE messages for computed column CASCADE drops.

Fixes #152841.

Release note (bug fix): Fixed deadlock in DROP COLUMN CASCADE operations when dropping columns referenced by STORED computed columns.

Epic: None

fqazi reviewed (approved):
Just some minor clean up suggestions the rules themselves look good.


<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @fqazi reviewed 2 of 3 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-:-O_XM4FG0yHQtBCqJWAG:b61nzx7)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_XL8po4Wx41H4l7rD1:-O_XL8po4Wx41H4l7rD2:bxys3yu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fb6f6866dead778b27564c66146644d8c10a02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L157)):*
> ```Go
> 				JoinOnDescID(from, to, "table-id"),
> 				JoinOnColumnID(from, colType, "table-id", "col-id"),
> 				FilterElements("computedColumnTypeReferencesColumn", from, to,
> ```

You can use the helper ReferencedColumnIDsContains instead
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_XLOg76dR_78FjJ8K-:-O_XLOg76dR_78FjJ8K0:b-cznxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fb6f6866dead778b27564c66146644d8c10a02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L168)):*
> ```Go
> 				FilterElements("columnTypeIsVirtual", colType, from,
> 					func(columnType *scpb.ColumnType, computeExpression *scpb.ColumnComputeExpression) bool {
> 						return columnType.IsVirtual && columnType.ColumnID == computeExpression.ColumnID
> ```

We can do a JoinOnColumnID for colType and columnExpression. 

 If we do think IsVirtual will be used more frequently, we can use rules and add it into attr.go. But I don't think that it is needed yet.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 209 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_XLuk4417TMBygU_pL:-O_XLuk4417TMBygU_pM:b1gpjpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fb6f6866dead778b27564c66146644d8c10a02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L209)):*
> ```Go
> 				FilterElements("columnTypeIsStored", colType, from,
> 					func(columnType *scpb.ColumnType, computeExpression *scpb.ColumnComputeExpression) bool {
> 						return !columnType.IsVirtual && columnType.ColumnID == computeExpression.ColumnID
> ```

Similar suggestion to use a join here.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153683)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_XL8po4Wx41H4l7rD1:-O_YD0l28Jgnmcz1Lnf3:baq2ci2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fb6f6866dead778b27564c66146644d8c10a02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L157)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

You can use the helper ReferencedColumnIDsContains instead
</blockquote></details>

ah right, I forgot about that one
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_XLOg76dR_78FjJ8K-:-O_YD3WD4BWNYkZ8MMyv:bfyjhzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fb6f6866dead778b27564c66146644d8c10a02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L168)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

We can do a JoinOnColumnID for colType and columnExpression. 

 If we do think IsVirtual will be used more frequently, we can use rules and add it into attr.go. But I don't think that it is needed yet.
</blockquote></details>

Good idea with the join. I'll hold off adding isVirtual as an attribute.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 209 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_XLuk4417TMBygU_pL:-O_YD7zJDEVZYkbT7mUV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fb6f6866dead778b27564c66146644d8c10a02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L209)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Similar suggestion to use a join here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: split computed column dependency rule to fix DROP COLUMN CASCADE deadlock

Previously, a single dependency rule applied to all computed column
expressions, causing deadlocks when dropping columns with STORED
computed dependencies via CASCADE. The rule forced computed expressions
to be dropped before their referenced columns, but this created circular
dependencies during CASCADE operations with STORED columns.

This change splits the rule into two separate rules:
- Virtual computed columns: Keep existing precedence rule needed for optimizer
- Stored computed columns: Only apply rule during ALTER COLUMN TYPE operations

This prevents the deadlock while preserving the necessary dependency ordering
for ALTER COLUMN TYPE operations where temporary expressions map old rows to
new column types.

Also adds NOTICE messages for computed column CASCADE drops.

Fixes #152841.

Release note (bug fix): Fixed deadlock in DROP COLUMN CASCADE operations
when dropping columns referenced by STORED computed columns.

Epic: None

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153683)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_XL8po4Wx41H4l7rD1:-O_qH5-V2DD5AMPgiXJ9:b62k7ll) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fb6f6866dead778b27564c66146644d8c10a02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L157)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

ah right, I forgot about that one
</blockquote></details>

I made this change, but for it to work I had to add referenced columns as a attribute to the ColumnComputeExpression. This ended up snowballing with all of the test changes.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_XLOg76dR_78FjJ8K-:-O_qJZ1A0NEiFSXOzvL9:b-lochn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fb6f6866dead778b27564c66146644d8c10a02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L168)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Good idea with the join. I'll hold off adding isVirtual as an attribute.
</blockquote></details>

I left the join out because we already do it higher up. The ColumnID check that I had in here was redundant.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 209 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_XLuk4417TMBygU_pL:-O_qJdXvCLxBAiM9jQq8:b-lochn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8fb6f6866dead778b27564c66146644d8c10a02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L209)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I left the join out because we already do it higher up. The ColumnID check that I had in here was redundant.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great work!

> @fqazi reviewed 136 of 136 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-:-O_qgrdlEdmBLjV8dWf0:bolbkdq)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 160 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_qgG-p2F32PF7qoE2Y:-O_qgG-p2F32PF7qoE2Z:b-8sf1pu) ([raw file](https://github.com/cockroachdb/cockroach/blob/35837b5e183b672fed0506e0acaf8250b81e0d9c/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L160)):*
> ```Go
> 				JoinOnColumnID(from, colType, "table-id", "target-col-id"),
> 				to.El.AttrEqVar(screl.ColumnID, colID),
> 				from.ReferencedColumnIDsContains(colID),
> ```

Nit: using target-col-id here is sufficient I think?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153683)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 160 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_qgG-p2F32PF7qoE2Y:-O_rOd7uD1j1-KqAtbf_:b-2gminn) ([raw file](https://github.com/cockroachdb/cockroach/blob/35837b5e183b672fed0506e0acaf8250b81e0d9c/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L160)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: using target-col-id here is sufficient I think?
</blockquote></details>

That didn't work for me. I had to use a separate var to save the column ID from `scpb.Column`. Otherwise, I end up getting plan differences. For instance,  in the end-to-end test `drop_multiple_columns_separate_statements`, some mutations show up at an earlier stage.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153683)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go` line 160 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/153683#-O_qgG-p2F32PF7qoE2Y:-O_r_LkbCzT0eupXEM3c:bqu6ozn) ([raw file](https://github.com/cockroachdb/cockroach/blob/35837b5e183b672fed0506e0acaf8250b81e0d9c/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_column.go#L160)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

That didn't work for me. I had to use a separate var to save the column ID from `scpb.Column`. Otherwise, I end up getting plan differences. For instance,  in the end-to-end test `drop_multiple_columns_separate_statements`, some mutations show up at an earlier stage.
</blockquote></details>

Okay, lets just leave it as-is then.


<!-- Sent from Reviewable.io -->
